File: crates/rocode-types/src/skill.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillSourceKind {
    Bundled,
    LocalPath,
    Git,
    Archive,
    Registry,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillSourceRef {
    pub source_id: String,
    pub source_kind: SkillSourceKind,
    pub locator: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub revision: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillSourceIndexEntry {
    pub skill_name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub revision: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub manifest_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub checksum: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillSourceIndexSnapshot {
    pub source: SkillSourceRef,
    pub updated_at: i64,
    #[serde(default)]
    pub entries: Vec<SkillSourceIndexEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillDistributionResolverKind {
    Bundled,
    LocalPath,
    RegistryIndex,
    RegistryManifest,
    ArchiveManifest,
    GitCheckout,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillArtifactKind {
    RegistryPackage,
    GitCheckout,
    Archive,
    LocalSnapshot,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillArtifactRef {
    pub artifact_id: String,
    pub kind: SkillArtifactKind,
    pub locator: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub checksum: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub size_bytes: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillDistributionRelease {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub revision: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub checksum: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub manifest_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub published_at: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillDistributionResolution {
    pub resolved_at: i64,
    pub resolver_kind: SkillDistributionResolverKind,
    pub artifact: SkillArtifactRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillInstalledDistribution {
    pub installed_at: i64,
    pub workspace_skill_path: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub installed_revision: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub local_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillManagedLifecycleState {
    Indexed,
    Resolved,
    Fetched,
    PlannedInstall,
    Installed,
    UpdateAvailable,
    Diverged,
    Detached,
    RemovePending,
    Removed,
    ResolutionFailed,
    FetchFailed,
    ApplyFailed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillDistributionRecord {
    pub distribution_id: String,
    pub source: SkillSourceRef,
    pub skill_name: String,
    pub release: SkillDistributionRelease,
    pub resolution: SkillDistributionResolution,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub installed: Option<SkillInstalledDistribution>,
    pub lifecycle: SkillManagedLifecycleState,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillArtifactCacheStatus {
    Fetched,
    Extracted,
    Failed,
    Evicted,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillArtifactCacheEntry {
    pub artifact: SkillArtifactRef,
    pub cached_at: i64,
    pub local_path: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub extracted_path: Option<String>,
    pub status: SkillArtifactCacheStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillManagedLifecycleRecord {
    pub distribution_id: String,
    pub source_id: String,
    pub skill_name: String,
    pub state: SkillManagedLifecycleState,
    pub updated_at: i64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct BundledSkillManifestEntry {
    pub skill_name: String,
    pub relative_path: String,
    pub content_hash: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct BundledSkillManifest {
    pub bundle_id: String,
    #[serde(default)]
    pub entries: Vec<BundledSkillManifestEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ManagedSkillRecord {
    pub skill_name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<SkillSourceRef>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub installed_revision: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub local_hash: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_synced_at: Option<i64>,
    #[serde(default)]
    pub locally_modified: bool,
    #[serde(default)]
    pub deleted_locally: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillSyncAction {
    Install,
    Update,
    SkipLocalModification,
    SkipDeletedLocally,
    RemoveManaged,
    Noop,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillSyncEntry {
    pub skill_name: String,
    pub action: SkillSyncAction,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillSyncPlan {
    pub source_id: String,
    #[serde(default)]
    pub entries: Vec<SkillSyncEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubManagedResponse {
    #[serde(default)]
    pub managed_skills: Vec<ManagedSkillRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubDistributionResponse {
    #[serde(default)]
    pub distributions: Vec<SkillDistributionRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubArtifactCacheResponse {
    #[serde(default)]
    pub artifact_cache: Vec<SkillArtifactCacheEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubPolicy {
    pub artifact_cache_retention_seconds: u64,
    pub fetch_timeout_ms: u64,
    pub max_download_bytes: u64,
    pub max_extract_bytes: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubPolicyResponse {
    pub policy: SkillHubPolicy,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubLifecycleResponse {
    #[serde(default)]
    pub lifecycle: Vec<SkillManagedLifecycleRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillRemoteInstallAction {
    Install,
    Update,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillRemoteInstallEntry {
    pub distribution_id: String,
    pub source_id: String,
    pub skill_name: String,
    pub action: SkillRemoteInstallAction,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillRemoteInstallPlan {
    pub source_id: String,
    pub distribution: SkillDistributionRecord,
    pub entry: SkillRemoteInstallEntry,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillRemoteInstallResponse {
    pub plan: SkillRemoteInstallPlan,
    pub artifact_cache: SkillArtifactCacheEntry,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub guard_report: Option<SkillGuardReport>,
    pub result: SkillGovernanceWriteResult,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubRemoteInstallPlanRequest {
    pub source: SkillSourceRef,
    pub skill_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubRemoteInstallApplyRequest {
    pub session_id: String,
    pub source: SkillSourceRef,
    pub skill_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubRemoteUpdatePlanRequest {
    pub source: SkillSourceRef,
    pub skill_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubRemoteUpdateApplyRequest {
    pub session_id: String,
    pub source: SkillSourceRef,
    pub skill_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillGovernanceWriteResult {
    pub action: String,
    pub skill_name: String,
    pub location: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub supporting_file: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubIndexResponse {
    #[serde(default)]
    pub source_indices: Vec<SkillSourceIndexSnapshot>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubIndexRefreshRequest {
    pub source: SkillSourceRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubIndexRefreshResponse {
    pub snapshot: SkillSourceIndexSnapshot,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubManagedDetachRequest {
    pub session_id: String,
    pub source: SkillSourceRef,
    pub skill_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubManagedDetachResponse {
    pub lifecycle: SkillManagedLifecycleRecord,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubManagedRemoveRequest {
    pub session_id: String,
    pub source: SkillSourceRef,
    pub skill_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubManagedRemoveResponse {
    pub lifecycle: SkillManagedLifecycleRecord,
    #[serde(default)]
    pub deleted_from_workspace: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<SkillGovernanceWriteResult>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubAuditResponse {
    #[serde(default)]
    pub audit_events: Vec<SkillAuditEvent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct SkillHubTimelineQuery {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub skill_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubTimelineResponse {
    #[serde(default)]
    pub entries: Vec<SkillGovernanceTimelineEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubSyncPlanRequest {
    pub source: SkillSourceRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubSyncApplyRequest {
    pub session_id: String,
    pub source: SkillSourceRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubSyncPlanResponse {
    pub plan: SkillSyncPlan,
    #[serde(default)]
    pub guard_reports: Vec<SkillGuardReport>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubGuardRunRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub skill_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<SkillSourceRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillHubGuardRunResponse {
    #[serde(default)]
    pub reports: Vec<SkillGuardReport>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillGovernanceTimelineKind {
    ManagedSnapshot,
    SourceIndexRefreshed,
    SourceResolved,
    ArtifactFetched,
    ArtifactEvicted,
    ArtifactFetchFailed,
    RemoteInstallPlanned,
    RemoteUpdatePlanned,
    LifecycleTransitioned,
    Create,
    Patch,
    Edit,
    Delete,
    WriteFile,
    RemoveFile,
    HubInstall,
    HubUpdate,
    HubDetach,
    HubRemove,
    SyncPlanCreated,
    SyncApplyCompleted,
    GuardBlocked,
    GuardWarned,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillGovernanceTimelineStatus {
    Info,
    Success,
    Warn,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillGovernanceTimelineEntry {
    pub entry_id: String,
    pub kind: SkillGovernanceTimelineKind,
    pub created_at: i64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub skill_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub actor: Option<String>,
    pub title: String,
    pub summary: String,
    pub status: SkillGovernanceTimelineStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub managed_record: Option<ManagedSkillRecord>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub guard_report: Option<SkillGuardReport>,
    #[serde(default)]
    pub payload: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillGuardSeverity {
    Info,
    Warn,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillGuardStatus {
    Passed,
    Warn,
    Blocked,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillGuardViolation {
    pub rule_id: String,
    pub severity: SkillGuardSeverity,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillGuardReport {
    pub skill_name: String,
    pub status: SkillGuardStatus,
    #[serde(default)]
    pub violations: Vec<SkillGuardViolation>,
    pub scanned_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillAuditKind {
    SourceIndexRefreshed,
    SourceResolved,
    ArtifactFetched,
    ArtifactEvicted,
    ArtifactFetchFailed,
    RemoteInstallPlanned,
    RemoteUpdatePlanned,
    LifecycleTransitioned,
    Create,
    Patch,
    Edit,
    Delete,
    WriteFile,
    RemoveFile,
    HubInstall,
    HubUpdate,
    HubDetach,
    HubRemove,
    SyncPlanCreated,
    SyncApplyCompleted,
    GuardBlocked,
    GuardWarned,
}

impl From<SkillAuditKind> for SkillGovernanceTimelineKind {
    fn from(value: SkillAuditKind) -> Self {
        match value {
            SkillAuditKind::SourceIndexRefreshed => Self::SourceIndexRefreshed,
            SkillAuditKind::SourceResolved => Self::SourceResolved,
            SkillAuditKind::ArtifactFetched => Self::ArtifactFetched,
            SkillAuditKind::ArtifactEvicted => Self::ArtifactEvicted,
            SkillAuditKind::ArtifactFetchFailed => Self::ArtifactFetchFailed,
            SkillAuditKind::RemoteInstallPlanned => Self::RemoteInstallPlanned,
            SkillAuditKind::RemoteUpdatePlanned => Self::RemoteUpdatePlanned,
            SkillAuditKind::LifecycleTransitioned => Self::LifecycleTransitioned,
            SkillAuditKind::Create => Self::Create,
            SkillAuditKind::Patch => Self::Patch,
            SkillAuditKind::Edit => Self::Edit,
            SkillAuditKind::Delete => Self::Delete,
            SkillAuditKind::WriteFile => Self::WriteFile,
            SkillAuditKind::RemoveFile => Self::RemoveFile,
            SkillAuditKind::HubInstall => Self::HubInstall,
            SkillAuditKind::HubUpdate => Self::HubUpdate,
            SkillAuditKind::HubDetach => Self::HubDetach,
            SkillAuditKind::HubRemove => Self::HubRemove,
            SkillAuditKind::SyncPlanCreated => Self::SyncPlanCreated,
            SkillAuditKind::SyncApplyCompleted => Self::SyncApplyCompleted,
            SkillAuditKind::GuardBlocked => Self::GuardBlocked,
            SkillAuditKind::GuardWarned => Self::GuardWarned,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkillAuditEvent {
    pub event_id: String,
    pub kind: SkillAuditKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub skill_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_id: Option<String>,
    pub actor: String,
    pub created_at: i64,
    #[serde(default)]
    pub payload: serde_json::Value,
}

File: crates/rocode-server/web/src/lib/skill.ts
export interface SkillCatalogEntry {
  name: string;
  description: string;
  category?: string | null;
  location: string;
  writable: boolean;
  supporting_files: string[];
}

export interface SkillFileRefRecord {
  relative_path: string;
  location: string;
}

export interface LoadedSkillMetaRecord {
  name: string;
  description: string;
  category?: string | null;
  location: string;
  supporting_files: SkillFileRefRecord[];
}

export interface LoadedSkillRecord {
  meta: LoadedSkillMetaRecord;
  content: string;
}

export interface SkillDetailResponseRecord {
  skill: LoadedSkillRecord;
  source: string;
  writable: boolean;
}

export interface SkillGovernanceWriteResultRecord {
  action: string;
  skill_name: string;
  location: string;
  supporting_file?: string | null;
}

export interface SkillMethodologyStepRecord {
  title: string;
  action: string;
  outcome?: string | null;
}

export interface SkillMethodologyReferenceRecord {
  label: string;
  path: string;
}

export interface SkillMethodologyTemplateRecord {
  when_to_use: string[];
  when_not_to_use: string[];
  prerequisites: string[];
  core_steps: SkillMethodologyStepRecord[];
  success_criteria: string[];
  validation: string[];
  pitfalls: string[];
  references: SkillMethodologyReferenceRecord[];
}

export interface SkillMethodologyPreviewResponseRecord {
  body: string;
}

export interface SkillMethodologyExtractResponseRecord {
  matched: boolean;
  methodology?: SkillMethodologyTemplateRecord | null;
}

export interface SkillGuardViolationRecord {
  rule_id: string;
  severity: "info" | "warn" | "error";
  message: string;
  file_path?: string | null;
}

export interface SkillGuardReportRecord {
  skill_name: string;
  status: "passed" | "warn" | "blocked";
  violations: SkillGuardViolationRecord[];
  scanned_at: number;
}

export interface SkillSourceRefRecord {
  source_id: string;
  source_kind: "bundled" | "local_path" | "git" | "archive" | "registry";
  locator: string;
  revision?: string | null;
}

export interface ManagedSkillRecord {
  skill_name: string;
  source?: SkillSourceRefRecord | null;
  installed_revision?: string | null;
  local_hash?: string | null;
  last_synced_at?: number | null;
  locally_modified: boolean;
  deleted_locally: boolean;
}

export interface SkillSourceIndexEntryRecord {
  skill_name: string;
  description?: string | null;
  category?: string | null;
  revision?: string | null;
}

export interface SkillSourceIndexSnapshotRecord {
  source: SkillSourceRefRecord;
  updated_at: number;
  entries: SkillSourceIndexEntryRecord[];
}

export interface SkillAuditEventRecord {
  event_id: string;
  kind: string;
  skill_name?: string | null;
  source_id?: string | null;
  actor: string;
  created_at: number;
  payload: unknown;
}

export interface SkillHubManagedResponseRecord {
  managed_skills: ManagedSkillRecord[];
}

export interface SkillHubIndexResponseRecord {
  source_indices: SkillSourceIndexSnapshotRecord[];
}

export interface SkillHubIndexRefreshResponseRecord {
  snapshot: SkillSourceIndexSnapshotRecord;
}

export interface SkillArtifactRefRecord {
  artifact_id: string;
  kind: string;
  locator: string;
  checksum?: string | null;
  size_bytes?: number | null;
}

export interface SkillDistributionReleaseRecord {
  version?: string | null;
  revision?: string | null;
  checksum?: string | null;
  manifest_path?: string | null;
  published_at?: number | null;
}

export interface SkillDistributionResolutionRecord {
  resolved_at: number;
  resolver_kind: string;
  artifact: SkillArtifactRefRecord;
}

export interface SkillInstalledDistributionRecord {
  installed_at: number;
  workspace_skill_path: string;
  installed_revision?: string | null;
  local_hash?: string | null;
}

export interface SkillDistributionRecord {
  distribution_id: string;
  source: SkillSourceRefRecord;
  skill_name: string;
  release: SkillDistributionReleaseRecord;
  resolution: SkillDistributionResolutionRecord;
  installed?: SkillInstalledDistributionRecord | null;
  lifecycle: string;
}

export interface SkillManagedLifecycleRecord {
  distribution_id: string;
  source_id: string;
  skill_name: string;
  state: string;
  updated_at: number;
  error?: string | null;
}

export interface SkillHubDistributionResponseRecord {
  distributions: SkillDistributionRecord[];
}

export interface SkillArtifactCacheEntryRecord {
  artifact: SkillArtifactRefRecord;
  cached_at: number;
  local_path: string;
  extracted_path?: string | null;
  status: string;
  error?: string | null;
}

export interface SkillHubArtifactCacheResponseRecord {
  artifact_cache: SkillArtifactCacheEntryRecord[];
}

export interface SkillHubPolicyRecord {
  artifact_cache_retention_seconds: number;
  fetch_timeout_ms: number;
  max_download_bytes: number;
  max_extract_bytes: number;
}

export interface SkillHubPolicyResponseRecord {
  policy: SkillHubPolicyRecord;
}

export interface SkillHubLifecycleResponseRecord {
  lifecycle: SkillManagedLifecycleRecord[];
}

export interface SkillHubAuditResponseRecord {
  audit_events: SkillAuditEventRecord[];
}

export interface SkillGovernanceTimelineEntryRecord {
  entry_id: string;
  kind: string;
  created_at: number;
  skill_name?: string | null;
  source_id?: string | null;
  actor?: string | null;
  title: string;
  summary: string;
  status: "info" | "success" | "warn" | "error";
  managed_record?: ManagedSkillRecord | null;
  guard_report?: SkillGuardReportRecord | null;
}

export interface SkillHubTimelineResponseRecord {
  entries: SkillGovernanceTimelineEntryRecord[];
}

export interface SkillSyncEntryRecord {
  skill_name: string;
  action: string;
  reason: string;
}

export interface SkillSyncPlanRecord {
  source_id: string;
  entries: SkillSyncEntryRecord[];
}

export interface SkillHubSyncPlanResponseRecord {
  plan: SkillSyncPlanRecord;
  guard_reports?: SkillGuardReportRecord[];
}

export interface SkillHubGuardRunRequestRecord {
  skill_name?: string;
  source?: SkillSourceRefRecord;
}

export interface SkillHubGuardRunResponseRecord {
  reports: SkillGuardReportRecord[];
}

export interface SkillRemoteInstallEntryRecord {
  distribution_id: string;
  source_id: string;
  skill_name: string;
  action: "install" | "update";
  reason: string;
}

export interface SkillRemoteInstallPlanRecord {
  source_id: string;
  distribution: SkillDistributionRecord;
  entry: SkillRemoteInstallEntryRecord;
}

export interface SkillManageResponseRecord {
  result: SkillGovernanceWriteResultRecord;
  guard_report?: SkillGuardReportRecord | null;
}

export interface SkillRemoteInstallResponseRecord {
  plan: SkillRemoteInstallPlanRecord;
  artifact_cache: SkillArtifactCacheEntryRecord;
  guard_report?: SkillGuardReportRecord | null;
  result: SkillGovernanceWriteResultRecord;
}

export interface SkillHubManagedDetachResponseRecord {
  lifecycle: SkillManagedLifecycleRecord;
}

export interface SkillHubManagedRemoveResponseRecord {
  lifecycle: SkillManagedLifecycleRecord;
  deleted_from_workspace: boolean;
  result?: SkillGovernanceWriteResultRecord | null;
}
